README.md
# digital_pay

This package allows you to make your payments safely. Designed to make your task easier with a simple widget.

## Features

 - `Paiment via MTN CI` 

## Getting started

To use this package, add digital_pay as a dependency in your pubspec.yaml file.

## Usage

Minimal example:

```dart
    //Launching the checkout request
    DigitalPay.checkout(
            context: context,
            accessToken: "MY_ACCESS_TOKEN",
            amount: 10)
        .then((result) {
          //result of the request
      print(result);
    });
```

Material App settings:

```dart
import 'dart:io';
import 'package:flutter/material.dart';

//Secure all hhtp request with [https]
class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(context) {
    return super.createHttpClient(context)
      ..badCertificateCallback =
          (X509Certificate cert, String host, int port) => true;
  }
}
//END secure http

//Change your main App
void main() async {
  //intialize http requests
  HttpOverrides.global = MyHttpOverrides();
  return runApp(const MyApp());
}
//END Change
```


CHANGELOG.md
## 1.0.0

* TODO: Initial release.

## 1.0.1

* TODO: Correction de Bugs.

## 1.0.2

* TODO: Migration iOS.

## 1.0.3

* TODO: Intégration de WAVE
* TODO: Intégration de VISA/MASTERCARD

example/test/widget_test.dart
import 'dart:io';

import 'package:digital_pay/digital_pay.dart';
import 'package:digital_pay/required.dart';
import 'package:flutter/material.dart';

//Secure all hhtp request with [https]
class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(context) {
    return super.createHttpClient(context)
      ..badCertificateCallback =
          (X509Certificate cert, String host, int port) => true;
  }
}
//END secure http

//Change your main App
void main() async {
  //intialize http requests
  HttpOverrides.global = MyHttpOverrides();
  return runApp(const MyApp());
}
//END Change

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'DIGITAL PAY',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("DIGITAL PAY"),
        centerTitle: true,
        backgroundColor: Colors.orange,
      ),
      body: Center(
        child: TextButton(
            style: ButtonStyle(
                backgroundColor: MaterialStateProperty.all(Colors.orange),
                padding: MaterialStateProperty.all(
                    const EdgeInsets.symmetric(horizontal: 50))),
            onPressed: () {
              DigitalPay.checkout(
                context: context,
                accessToken: "MY_ACCESS_TOKEN",
                amount: 10000,
                color: Colors.red
              ).then((result) {
                // ignore: avoid_print
                (result);
                toastMsg(result["message"]);
              });
            },
            child: const Text(
              "Payer",
              style: TextStyle(color: Colors.white),
            )),
      ),
    );
  }
}


lib/required.dart
import 'dart:convert';

import 'package:currency_formatter/currency_formatter.dart';
import 'package:digital_pay/constant.dart';
import 'package:flutter/material.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';
import 'package:fluttertoast/fluttertoast.dart';
import 'package:http/http.dart' as http;
import 'package:mask_text_input_formatter/mask_text_input_formatter.dart';

void unloading(BuildContext context, [Map? data]) {
  Navigator.pop(context, data);
}

loading(context) {
  showDialog(
      context: context,
      builder: (c) {
        return WillPopScope(
          onWillPop: () {
            return Future.value(true);
          },
          child: Scaffold(
            backgroundColor: Colors.transparent,
            body: Center(
              child: Container(
                width: 180,
                height: 80,
                decoration: BoxDecoration(
                  color: Colors.white,
                  border:
                      Border(top: BorderSide(color: primaryColor, width: 5)),
                ),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    SpinKitCircle(
                      color: primaryColor,
                      size: 40,
                    ),
                    const Text("Chargement...")
                  ],
                ),
              ),
            ),
          ),
        );
      });
}

toastMsg(String msg, [String? type = "danger"]) {
  Fluttertoast.showToast(
      msg: msg,
      backgroundColor: type == "danger" ? Colors.red : Colors.green,
      gravity: ToastGravity.TOP,
      toastLength: Toast.LENGTH_LONG);
}

class ApiDCI {
  static Future get(uri, path, [Map? params, String? _token]) async {
    var q = "";
    var link = "";
    path = "/" + path;
    if (params != null) {
      q = "?";

      params.forEach((key, value) {
        q += key.toString() + "=" + value.toString() + "&";
      });
      _token != null ? q += "_token=$_token" : q = q.substring(0, q.length - 1);
      link = uri + path + q;
    } else {
      link = uri + path;
    }
    try {
      var response = await http.get(Uri.parse(link), headers: _setHeaders());
      var result = response.body;
      if (response.statusCode == 200 && result != "error") {
        return jsonDecode(result);
      } else {
        return "error";
      }
    } catch (e) {
      return "error";
    }
  }

  static Future gestSellerInfo(String accessToken) async {
    var response = await get(apiLink, "seller", {"code": accessToken});
    return response;
  }

  static Future paiement({
    required int amount,
    required String accessToken,
    required String number,
    required String operator,
  }) async {
    var response = await get(apiLink, "paiement", {
      "access_token": accessToken,
      "amount": amount,
      "number": "225" + number,
      "operator": operator,
    });
    return response;
  }

  static Future paiementWave({
    required int amount,
    required String accessToken,
  }) async {
    var response = await get("$apiLink/wave", "paiement", {
      "amount": amount,
      "access_token": accessToken,
    });
    return response;
  }

  static Future verificationWave({required String id}) async {
    var response = await get("$apiLink/wave", "verification", {"id": id});
    return response;
  }

  static Future verification({required String accessToken}) async {
    var response =
        await get(apiLink, "verification", {"access_token": accessToken});
    return response;
  }

  static Future<bool> setErrorTransaction(String idTransaction) async {
    try {
      return await get(apiLink, "seterror", {"id_transaction": idTransaction});
    } catch (e) {
      return false;
    }
  }

  static _setHeaders() => {
        'Content-type': 'application/json',
        'Accept': 'application/json',
      };
}

var maskNumero = MaskTextInputFormatter(
  mask: '##-####-####',
  filter: {
    "#": RegExp(r'[0-9]'),
  },
);

String unMask(String string) {
  string = string.replaceAll("-", "");
  return string.trim();
}

String numberFormat(int amount) {
  CurrencyFormatterSettings euroSettings = CurrencyFormatterSettings(
    symbol: 'FCFA',
    symbolSide: SymbolSide.right,
    thousandSeparator: ' ',
    symbolSeparator: ' ',
  );

  return CurrencyFormatter.format(amount, euroSettings, decimal: 0);
}


lib/digital_pay.dart
library digital_pay;

import 'dart:async';
import 'package:digital_pay/components/momo.dart';
import 'package:digital_pay/components/visa.dart';
import 'package:digital_pay/components/wave.dart';
import 'package:digital_pay/required.dart';
import 'package:flutter/material.dart';

import 'constant.dart';

class DigitalPay {
  /// Payer avec le package [DigitalPay] en toute sécurité.<br/>
  /// ### Obtenez votre code d'accès [accessToken] via https://pay.digital.ci
  /// #### ```Le montant de la transaction [amount] doit être supérieur à 10.```
  /// #### ** Codes de retour
  /// #### ----------------------------------
  /// ##### "INTERNAL_ERROR" : Erreur interne
  /// ``` veuillez contacter le support: ``` https://pay.digital.ci
  /// ##### "NOT_PERMITED" : Utlisateur non autorisé
  /// ##### "SERVER_ERROR" : Erreur du serveur
  /// ##### "TIME_OFF" : Temps du delai écoulé
  /// ##### "FAIL_TRANSACTION" : transaction échouée
  /// ##### "SUCCESS_TRANSACTION" : Transaction  effectuée
  /// ##### "CANCEL" : Opération annulée

  static Future<Map> checkout({
    required BuildContext context,
    required accessToken,
    required amount,
    bool appBar = true,
    Color color = Colors.transparent,
  }) async {
    try {
      Map result =
          await Navigator.push(context, MaterialPageRoute(builder: (c) {
        return HomeDPay(
          accessToken: accessToken,
          amount: amount,
          appBar: appBar,
          color: color,
        );
      }));
      return result;
    } catch (e) {
      return {
        "code": "INTERNAL_ERROR",
        "message": "Désoler une erreur s'est produite!!!"
      };
    }
  }

  static Future verification({required String transactionID}) async {
    return await ApiDCI.verification(accessToken: transactionID);
  }
}

//Payer en toute sécurité avec notre système
class HomeDPay extends StatefulWidget {
  /// Obtenir votre accessToken [accessToken] via https://pay.digital.ci
  final String accessToken;
  final bool appBar;

  /// Le montant [amount] de la transaction doit être supérieur à 10
  final int amount;

  /// La couleur du formulaire [color]
  final Color color;

  final Function(dynamic)? waitResponse;
  const HomeDPay({
    Key? key,
    required this.accessToken,
    required this.amount,
    required this.appBar,
    this.waitResponse,
    required this.color,
  }) : super(key: key);

  @override
  State<HomeDPay> createState() => _HomeDPayState();
}

class _HomeDPayState extends State<HomeDPay> {
  Map assets = {
    "momo": [
      "https://raw.githubusercontent.com/KBenjaminYao/digital_pay/main/assets/momo.png",
      "MTN",
      true
    ],
    "wave": [
      "https://raw.githubusercontent.com/KBenjaminYao/digital_pay/main/assets/wave.png",
      "WAVE",
      false
    ],
    // "moov": [
    //   "https://raw.githubusercontent.com/KBenjaminYao/digital_pay/main/assets/moov.png",
    //   "MOOV",
    //   false
    // ],
    // "orange": [
    //   "https://raw.githubusercontent.com/KBenjaminYao/digital_pay/main/assets/orange.png",
    //   "OM",
    //   false
    // ],
    "visa": [
      "https://raw.githubusercontent.com/KBenjaminYao/digital_pay/main/assets/visa.png",
      "VISA",
      false
    ]
  };
  String phoneCode = "+225";
  String operator = "MTN";
  TextEditingController contactController = TextEditingController();
  final formKey = GlobalKey<FormState>();
  final maxTime = 180;
  bool isReady = true;
  bool? failTransaction;
  late int counter;
  Timer? _timer;
  Map userInfo = {};
  String idTransaction = "";
  Widget data = Container();

  @override
  void initState() {
    counter = maxTime;
    formRouter("momo");
    Future.delayed(Duration.zero).then((value) {
      loading(context);
      ApiDCI.gestSellerInfo(widget.accessToken).then((value) {
        unloading(context);
        if (value != "error") {
          if (value.isNotEmpty) {
            setState(() {
              userInfo = value;
              primaryColor = widget.color;
            });
          } else {
            if (mounted) {
              String message =
                  "Désoler vous n'êtes pas autorisé à utiliser ce service!!!";
              unloading(context, {"code": "NOT_PERMITED", "message": message});
            }
          }
        } else {
          if (mounted) {
            String message = "Désoler une erreur s'est produite!!!";
            unloading(context, {"code": "SERVER_ERROR", "message": message});
          }
        }
      });
    });
    super.initState();
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    if (counter == 0) {
      _timer?.cancel();
      ApiDCI.setErrorTransaction(idTransaction).then((statut) {
        if (mounted) {
          setState(() {
            counter = maxTime;
            failTransaction = statut;
          });
        }
        if (statut) {
          if (mounted) {
            setState(() {
              isReady = true;
            });
            String message = "Temps de la transaction écoulé!!!";
            unloading(context, {"code": "TIME_OFF", "message": message});
          }
        } else {
          if (mounted) {
            String message = "Désoler une erreur s'est produite!!!";
            unloading(context, {"code": "SERVER_ERROR", "message": message});
          }
        }
      });
    } else {}
    return GestureDetector(
      onTap: () {
        FocusScope.of(context).requestFocus(FocusNode());
      },
      child: WillPopScope(
          child: Scaffold(
            backgroundColor: Colors.white,
            appBar: widget.appBar
                ? AppBar(
                    backgroundColor: primaryColor,
                    title: Text(userInfo["name"] ?? ""),
                    centerTitle: true,
                  )
                : null,
            body: Padding(
              padding: const EdgeInsets.symmetric(vertical: 10),
              child: ListView(
                physics: const BouncingScrollPhysics(),
                children: [
                  SizedBox(
                    height: 80,
                    width: MediaQuery.of(context).size.width,
                    child: Center(
                      child: Container(
                        padding: const EdgeInsets.only(bottom: 15),
                        child: ListView(
                          shrinkWrap: true,
                          scrollDirection: Axis.horizontal,
                          children: [
                            buttonPaiement(assets["momo"][0], assets["momo"][1],
                                "momo", assets["momo"][2], false),
                            buttonPaiement(assets["wave"][0], assets["wave"][1],
                                "wave", assets["wave"][2], false),
                            buttonPaiement(assets["visa"][0], assets["visa"][1],
                                "visa", assets["visa"][2], false),
                          ],
                        ),
                      ),
                    ),
                  ),
                  Container(
                    alignment: Alignment.center,
                    child: const Text(
                      "DIGITAL PAY",
                      style:
                          TextStyle(fontWeight: FontWeight.bold, fontSize: 20),
                    ),
                  ),
                  Divider(
                    indent: 20,
                    endIndent: 20,
                    thickness: 2,
                    color: Colors.orange.withOpacity(.5),
                  ),
                  const SizedBox(height: 10),
                  Padding(
                    padding: const EdgeInsets.symmetric(horizontal: 20.0),
                    child: Text(
                      numberFormat(widget.amount),
                      textAlign: TextAlign.end,
                      style: const TextStyle(
                          fontWeight: FontWeight.bold, fontSize: 25),
                    ),
                  ),
                  const SizedBox(height: 5),
                  // MomoWidget(
                  //   accessToken: widget.accessToken,
                  //   amount: widget.amount,
                  // ),
                  data,
                ],
              ),
            ),
          ),
          onWillPop: () {
            unloading(
                context, {"code": "CANCEL", "message": "Opération annulée"});
            return Future.value(false);
          }),
    );
  }

  Widget buttonPaiement(String image, String op, String code, bool active,
      [bool? desactived = true]) {
    return Stack(
      clipBehavior: Clip.none,
      children: [
        GestureDetector(
          onTap: desactived!
              ? null
              : () {
                  setState(() {
                    operator = op;
                    assets["momo"][2] = false;
                    assets["visa"][2] = false;
                    assets["wave"][2] = false;
                    assets[code][2] = true;
                  });

                  formRouter(code);
                },
          child: Opacity(
            opacity: active ? 1 : .4,
            child: Card(
              elevation: 5,
              child: Container(
                width: 70,
                margin: const EdgeInsets.all(5),
                decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(10),
                    image: DecorationImage(
                      image: NetworkImage(image),
                      fit: BoxFit.fill,
                    ),
                    border: Border.all(
                        color: Colors.grey.withOpacity(.5), width: 2)),
              ),
            ),
          ),
        ),
        active
            ? Positioned(
                bottom: -12,
                child: SizedBox(
                  width: 90,
                  child: Center(
                    child: CircleAvatar(
                      radius: 6,
                      backgroundColor: primaryColor,
                    ),
                  ),
                ),
              )
            : Container()
      ],
    );
  }

  formRouter(String code) {
    setState(() {
      if (code == "wave") {
        data = WaveWidget(
          accessToken: widget.accessToken,
          amount: widget.amount,
        );
      }
      if (code == "momo") {
        data = MomoWidget(
          accessToken: widget.accessToken,
          amount: widget.amount,
        );
      }
      if (code == "visa") {
        data = VisaWidget(
          accessToken: widget.accessToken,
          amount: widget.amount,
        );
      }
    });
  }
}


lib/components/momo.dart
import 'dart:async';

import 'package:country_picker/country_picker.dart';
import 'package:digital_pay/constant.dart';
import 'package:digital_pay/required.dart';
import 'package:flutter/material.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';
import 'package:wc_form_validators/wc_form_validators.dart';

class MomoWidget extends StatefulWidget {
  final int amount;
  final String accessToken;
  const MomoWidget({Key? key, required this.amount, required this.accessToken})
      : super(key: key);

  @override
  State<MomoWidget> createState() => _MomoWidgetState();
}

class _MomoWidgetState extends State<MomoWidget> {
  String phoneCode = "+225";
  String operator = "MTN";
  TextEditingController contactController = TextEditingController();
  final formKey = GlobalKey<FormState>();
  final maxTime = 180;
  bool isReady = true;
  bool? failTransaction;
  late int counter;
  Timer? _timer;
  Map userInfo = {};
  String idTransaction = "";

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 20.0),
      child: Form(
        key: formKey,
        child: Column(
          children: [
            Container(
              padding: const EdgeInsets.only(bottom: 10),
              decoration: BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.circular(10.0),
                boxShadow: [
                  BoxShadow(
                    blurRadius: 1.0,
                    spreadRadius: 1.0,
                    color: secondaryColor,
                  ),
                ],
              ),
              child: ListTile(
                leading: Container(
                  padding: const EdgeInsets.only(top: 2),
                  child: GestureDetector(
                      onTap: () {
                        showCountryPicker(
                          context: context,
                          countryFilter: ["CI"],
                          countryListTheme: CountryListThemeData(
                            flagSize: 25,
                            backgroundColor: Colors.white,
                            textStyle: const TextStyle(
                                fontSize: 16, color: Colors.blueGrey),
                            bottomSheetHeight:
                                250, // Optional. Country list modal height
                            //Optional. Sets the border radius for the bottomsheet.
                            borderRadius: const BorderRadius.only(
                              topLeft: Radius.circular(20.0),
                              topRight: Radius.circular(20.0),
                            ),
                            //Optional. Styles the search field.
                            inputDecoration: InputDecoration(
                              labelText: 'Rechercher',
                              hintText: 'Recherchez un pays',
                              prefixIcon: const Icon(Icons.search),
                              border: OutlineInputBorder(
                                borderSide: BorderSide(
                                  color:
                                      const Color(0xFF8C98A8).withOpacity(0.2),
                                ),
                              ),
                            ),
                          ),
                          onSelect: (Country country) {
                            setState(() {
                              phoneCode = "+" + country.phoneCode;
                            });
                          },
                        );
                      },
                      child: Container(
                        padding: const EdgeInsets.only(top: 3),
                        child: Text(
                          phoneCode,
                          style: TextStyle(
                              fontSize: 14,
                              fontWeight: FontWeight.bold,
                              color: primaryColor),
                        ),
                      )),
                ),
                title: TextFormField(
                  enabled: isReady,
                  validator: Validators.compose([
                    Validators.required("Le numéro est Oligatoire"),
                    Validators.minLength(
                        12, "Veuillez renseigner un numéro correct"),
                    Validators.minLength(
                        12, "Veuillez renseigner un numéro correct"),
                    Validators.patternRegExp(RegExp(RegExp.escape("05")),
                        "Numéro $operator incorrect")
                  ]),
                  controller: contactController,
                  inputFormatters: [maskNumero],
                  style: TextStyle(
                      color: primaryColor, fontWeight: FontWeight.bold),
                  keyboardType: TextInputType.phone,
                  decoration: const InputDecoration(
                    hintText: 'Numéro de téléphone',
                    border: InputBorder.none,
                    errorStyle: TextStyle(
                      fontSize: 12,
                      height: .06,
                    ),
                  ),
                ),
              ),
            ),
            const SizedBox(
              height: 5,
            ),
            const Center(
              child: Text(
                "Sécurisé avec DIGITAL PAY",
                style: TextStyle(color: Colors.grey),
              ),
            ),
            isReady
                ? Container()
                : Column(
                    children: [
                      ListTile(
                        title: Text(
                          failTransaction == null
                              ? "Veuillez composer *133# choisir puis  l'option 1 pour confirmer la transaction"
                              : "Nous n'avons pas pu procéder à la vérification de votre transaction",
                          style: TextStyle(
                              color: failTransaction == null
                                  ? Colors.black
                                  : Colors.red),
                          textAlign: TextAlign.center,
                        ),
                        trailing: failTransaction != null
                            ? null
                            : Container(
                                width: 50,
                                height: 50,
                                alignment: Alignment.center,
                                padding: const EdgeInsets.all(5),
                                decoration: BoxDecoration(
                                    border: Border.all(
                                        width: 3, color: primaryColor),
                                    borderRadius: BorderRadius.circular(200)),
                                child: Text(
                                  counter < 10 ? "0$counter" : "$counter",
                                  style: const TextStyle(
                                    fontSize: 16,
                                    fontWeight: FontWeight.bold,
                                  ),
                                  textAlign: TextAlign.center,
                                ),
                              ),
                      ),
                      failTransaction != null
                          ? Container()
                          : SpinKitWave(
                              color: secondaryColor,
                              size: 20,
                            ),
                    ],
                  ),
            Container(
                width: MediaQuery.of(context).size.width - 60,
                margin: const EdgeInsets.only(top: 10),
                child: ElevatedButton(
                    style: ButtonStyle(
                        backgroundColor:
                            MaterialStateProperty.all(primaryColor)),
                    onPressed: !isReady
                        ? null
                        : () async {
                            bool state = formKey.currentState!.validate();
                            if (!state) {
                              return;
                            }
                            if (failTransaction != null) {
                              starTimer(idTransaction);
                            } else {
                              momoPaiementAction();
                            }
                          },
                    child:
                        Text((failTransaction != null) ? "Ressayer" : "PAYER")))
          ],
        ),
      ),
    );
  }

  void starTimer(String idTransaction) {
    counter = maxTime;
    _timer = Timer.periodic(const Duration(seconds: 1), (timer) async {
      if (mounted) {
        setState(() {
          counter--;
        });
      }
      await verificationTrans(idTransaction);
    });
  }

  verificationTrans(String idTransaction) async {
    await ApiDCI.verification(accessToken: idTransaction).then((value) {
      if (value != "error") {
        if (value["status"] == "FAILED") {
          _timer?.cancel();
          Map result = {
            "code": "FAIL_TRANSACTION",
            "message": "Transaction échouée!!!",
            "id": idTransaction
          };
          result.addAll(value);
          if (mounted) {
            unloading(context, result);
          }
        } else if (value["status"] == "SUCCESSFUL") {
          String message = "Transaction  effectuée !!!";
          Map result = {
            "code": "SUCCESS_TRANSACTION",
            "message": message,
            "id": idTransaction
          };
          result.addAll(value);
          if (mounted) {
            unloading(context, result);
          }
        }
      } else {
        if (mounted) {
          String message = "Désoler une erreur s'est produite!!!";
          unloading(context, {"code": "SERVER_ERROR", "message": message});
        }
      }
    });
  }

  void momoPaiementAction() {
    loading(context);
    ApiDCI.paiement(
      amount: widget.amount,
      accessToken: widget.accessToken,
      number: unMask(contactController.text),
      operator: operator,
    ).then((value) {
      unloading(context);
      if (value != "error") {
        Map data = {};
        try {
          data = value;
          setState(() {
            idTransaction = data["code"];
            isReady = false;
            starTimer(idTransaction);
          });
        } catch (e) {
          setState(() {
            isReady = true;
            counter = maxTime;
          });
          toastMsg("Impossible de poursuivre l'opération! Veuillez réessayer.");
        }
      } else {
        setState(() {
          isReady = true;
          counter = maxTime;
        });

        toastMsg("Opération echouée! Veuillez réessayer.");
      }
    });
  }
}


lib/components/visa.dart
import 'dart:convert';

import 'package:digital_pay/constant.dart';
import 'package:digital_pay/required.dart';
import 'package:flutter/material.dart';
import 'package:flutter_inappwebview/flutter_inappwebview.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';
import 'package:html/parser.dart';

class VisaWidget extends StatefulWidget {
  final String accessToken;
  final int amount;
  const VisaWidget({Key? key, required this.accessToken, required this.amount})
      : super(key: key);

  @override
  State<VisaWidget> createState() => _VisaWidgetState();
}

class _VisaWidgetState extends State<VisaWidget> {
  bool isReady = false;
  InAppWebViewController? webViewController;
  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: MediaQuery.of(context).size.height * .5,
      child: Stack(
        children: [
          InAppWebView(
            onLoadStop: (controller, url) async {
              setState(() {
                isReady = true;
              });
              String uri = url!.host + url.path;
              if (uri == host) {
                var data = await webViewController!.getHtml();
                var document = parse(data);
                final render = parse(document.body!.text).documentElement!.text;
                try {
                  Map reponse = jsonDecode(render);

                  if (reponse["statut"] == "error") {
                    Map result = {
                      "code": "FAIL_TRANSACTION",
                      "message": "Transaction échouée!!!",
                      "gt_msg": reponse["gtpay_tranx_status_msg"]
                    };
                    unloading(context, result);
                  } else {
                    Map result = {
                      "code": "SUCCESS_TRANSACTION",
                      "message": "Transaction effectuée!!!",
                    };
                    result.addAll(reponse);
                    unloading(context, result);
                  }
                } catch (e) {
                  return Future.error(e);
                }
              }
            },
            onWebViewCreated: (controller) {
              webViewController = controller;
            },
            onLoadStart: (controller, url) {
              setState(() {
                isReady = false;
              });
            },
            onReceivedServerTrustAuthRequest: (controller, challenge) async {
              return ServerTrustAuthResponse(
                  action: ServerTrustAuthResponseAction.PROCEED);
            },
            initialUrlRequest: URLRequest(
              url: Uri.https("pay.digital.ci", "api/gtb/form", {
                "access_token": widget.accessToken,
                "amount": widget.amount.toString(),
              }),
            ),
          ),
          !isReady
              ? Positioned(
                  top: 0,
                  left: 20,
                  child: Container(
                    margin: const EdgeInsets.only(top: 20),
                    width: MediaQuery.of(context).size.width - 50,
                    child: SpinKitPouringHourGlassRefined(
                      color: primaryColor,
                      size: 80,
                      strokeWidth: 1,
                    ),
                  ))
              : Container()
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'dart:io';

import 'package:digital_pay/digital_pay.dart';
import 'package:digital_pay/required.dart';
import 'package:flutter/material.dart';

//Secure all hhtp request with [https]
class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(context) {
    return super.createHttpClient(context)
      ..badCertificateCallback =
          (X509Certificate cert, String host, int port) => true;
  }
}
//END secure http

//Change your main App
void main() async {
  //intialize http requests
  HttpOverrides.global = MyHttpOverrides();
  return runApp(const MyApp());
}
//END Change

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'DIGITAL PAY',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("DIGITAL PAY"),
        centerTitle: true,
        backgroundColor: Colors.orange,
      ),
      body: Center(
        child: TextButton(
            style: ButtonStyle(
                backgroundColor: MaterialStateProperty.all(Colors.orange),
                padding: MaterialStateProperty.all(
                    const EdgeInsets.symmetric(horizontal: 50))),
            onPressed: () {
              DigitalPay.checkout(
                      context: context,
                      accessToken: "MY_ACCESS_TOKEN",
                      amount: 10000,
                      color: Colors.red)
                  .then((result) {
                // ignore: avoid_print
                (result);
                toastMsg(result["message"]);
              });
            },
            child: const Text(
              "Payer",
              style: TextStyle(color: Colors.white),
            )),
      ),
    );
  }
}
